pennylane/wires.py

```diff
@@ -0,0 +1,399 @@
+    which are just wire registers of length one.
+
+    Indexing and slicing this sequence will return another ``Wires`` object.
+
+    Args:
+         wires (Any): If iterable, interpreted as an ordered collection of unique objects representing wires.
+            Any other input type is converted into an iterable of a single entry, and hence interpreted as a
+            single wire.
+    """
+
+    def __init__(self, wires):
+
+        if isinstance(wires, Wires):
+            # If input is already a Wires object, just copy its wire list
+            self.wire_list = wires.wire_list
+
+        elif isinstance(wires, Iterable):
+            # If input is an iterable, check that entries are unique and convert to a list
+            try:
+                if len(set(wires)) != len(wires):
+                    raise WireError(
+                        "Wires must be unique; got {}.".format(wires)
+                    )
+            except TypeError:
+                raise WireError("Cannot create wires from elements of type {}.".format(type(wires[0]))) # TODO: edge case that iterable contains different types
+
+            self.wire_list = list(wires)
+
+        else:
+            # If input is not an iterable, interpret it as an object representing a single wire
+            self.wire_list = [wires]
+
+    def __getitem__(self, idx):
+        """Method to support indexing. Returns a Wires object representing a register with a single wire."""
+        return Wires(self.wire_list[idx])
+
+    def __len__(self):
+        """Method to support ``len()``."""
+        return len(self.wire_list)
+
+    def __contains__(self, item):
+        """Method checking if Wires object contains an object."""
+        if isinstance(item, Wires):
+            # If all wires can be found in this object, return True
+            if all(wire in self.wire_list for wire in item.wire_list):
+                return True
+        return False
+
+    def __repr__(self):
+        """Method defining the string representation of this class."""
+        return "<Wires = {}>".format(self.wire_list)
+
+    def __eq__(self, other):
+        """Method to support the '==' operator. This will also implicitly define the '!=' operator."""
+        if isinstance(other, self.__class__):
+            return self.wire_list == other.wire_list
+        return False
+
+    def __hash__(self):
+        """Implements the hash function, used for example by ``set()``."""
+        # hash the string representation
+        return hash(str(repr(self)))
+
+    def as_ndarray(self):
+        """Returns a numpy array representation of the Wires object.
+
+        Returns:
+            ndarray: array representing Wires object
+        """
+        return np.array(self.wire_list)
+
+    def as_list(self):
+        """Returns a list representation of the Wires object.
+
+        Returns:
+            List: list representing Wires object
+        """
+        return self.wire_list
```

Comment:
Minor nitpick: better to name this method to `tolist()`?

* `tolist()` is already used in NumPy/Pandas etc., so it's the one that feels more intuitive to me.

* In NumPy, `assomething()` _tends_ to be used for methods that don't copy data, and instead create a new view into that data. Examples include `np.asarray()` --- the object type is modified, but the underlying data remains in the same memory location.

  `tosomething() is instead used when there is implicit copying of the underlying data happening.

Nitpick because I don't think even NumPy is consistent here, `np.astype()` creates a copy by default iirc 😆

Reply:
It's shorter, so I like the idea.